PySDM/dynamics/collisions/collision.py

```diff
@@ -0,0 +1,298 @@
+        self.n_fragment = self.particulator.PairwiseStorage.empty(**empty_args_pairwise)
+        self.Ec_temp = self.particulator.PairwiseStorage.empty(**empty_args_pairwise)
+        self.Eb_temp = self.particulator.PairwiseStorage.empty(**empty_args_pairwise)
+        neg_ones_tmp = np.tile([-1], self.particulator.n_sd // 2)
+        self.neg_ones = self.particulator.PairwiseStorage.from_ndarray(neg_ones_tmp)
+        self.norm_factor_temp = self.particulator.Storage.empty(**empty_args_cellwise)
+        self.prob = self.particulator.PairwiseStorage.empty(**empty_args_pairwise)
+        self.is_first_in_pair = self.particulator.PairIndicator(self.particulator.n_sd)
+        self.dt_left = self.particulator.Storage.empty(**empty_args_cellwise)
+
+        self.stats_n_substep = self.particulator.Storage.empty(
+            self.particulator.mesh.n_cell, dtype=int)
+        self.stats_n_substep[:] = 0 if self.adaptive else self.__substeps
+        self.stats_dt_min = self.particulator.Storage.empty(**empty_args_cellwise)
+        self.stats_dt_min[:] = np.nan
+
+        self.rnd_opt_coll.register(builder)
+        self.rnd_opt_proc.register(builder)
+        self.rnd_opt_frag.register(builder)
+        self.kernel.register(builder)
+        self.coal_eff.register(builder)
+        self.break_eff.register(builder)
+        self.fragmentation.register(builder)
+
+        if self.croupier is None:
+            self.croupier = self.particulator.backend.default_croupier
+
+        self.collision_rate = self.particulator.Storage.from_ndarray(
+            np.zeros(self.particulator.mesh.n_cell, dtype=int))
+        self.collision_rate_deficit = self.particulator.Storage.from_ndarray(
+            np.zeros(self.particulator.mesh.n_cell, dtype=int))
+        self.coalescence_rate = self.particulator.Storage.from_ndarray(
+            np.zeros(self.particulator.mesh.n_cell, dtype=int))
+        self.breakup_rate = self.particulator.Storage.from_ndarray(
+            np.zeros(self.particulator.mesh.n_cell, dtype=int))
+
+    def __call__(self):
+        if self.enable:
+            if not self.adaptive:
+                for _ in range(self.__substeps):
+                    self.step()
+            else:
+                self.dt_left[:] = self.particulator.dt
+
+                while self.particulator.attributes.get_working_length() != 0:
+                    self.particulator.attributes.cell_idx.sort_by_key(self.dt_left)
+                    self.step()
+
+                self.particulator.attributes.reset_working_length()
+                self.particulator.attributes.reset_cell_idx()
+            self.rnd_opt_coll.reset()
+            self.rnd_opt_proc.reset()
+            self.rnd_opt_frag.reset()
+
+    def step(self):
+        # (1) Make the superdroplet list and random numbers for collision, process, fragmentation
+        pairs_rand, rand = self.rnd_opt_coll.get_random_arrays()
+        proc_rand = self.rnd_opt_proc.get_random_arrays()
```

Comment:
@edejong-caltech IIUC, we'd best avoid generating random numbers for proc_rand and rand_frag in case only coalescence is selected?

Reply:
I agree, but I'm not sure there is  a good way to have Collision be a superclass without doing this...

Reply:
I'll work on refactoring it (probably in a similar way as advection+sedimentation is handled and enabled/disabled in Displacement dynamic)